transports/sslsocket.go

```diff
@@ -1,27 +1,139 @@
 package transports
 
 import (
+	"compress/zlib"
 	"crypto/tls"
 	"fmt"
+	"io"
 	"net"
+	"time"
 )
 
 type sslsocket struct {
 	net.Conn
 }
 
+type zlibConn struct {
+	rawConn          net.Conn
+	compressionLevel int
+
+	zw *zlib.Writer
+	zr io.ReadCloser
+}
+
+func newZlibConn(rawConn net.Conn, compressionLevel int) (*zlibConn, error) {
+	conn := &zlibConn{
+		rawConn:          rawConn,
+		compressionLevel: compressionLevel,
+	}
+
+	return conn, nil
+}
+
+var _ net.Conn = new(zlibConn)
+
+func (c *zlibConn) Read(b []byte) (n int, err error) {
+
+	if c.zr == nil {
+		c.zr, err = zlib.NewReader(c.rawConn)
+		if err != nil {
+			return 0, fmt.Errorf("zlibconn.read. zlib.NewReader. %w", err)
+		}
+	}
+
+	return c.zr.Read(b)
+}
+
+func (c *zlibConn) Write(b []byte) (n int, err error) {
+
+	if c.zw == nil {
+		c.zw, err = zlib.NewWriterLevel(c.rawConn, c.compressionLevel)
+		if err != nil {
+			return 0, fmt.Errorf("zlibconn.write. zlib.NewWriterLevel. %w", err)
+		}
+	}
+
+	n, err = c.zw.Write(b)
+	if err != nil {
+		return 0, fmt.Errorf("zlibconn.write err: %w", err)
+	}
+
+	return n, c.zw.Flush()
+}
+
+func (c *zlibConn) Close() error {
+	if c.zr != nil {
+		if err := c.zr.Close(); err != nil {
+			return err
+		}
+	}
+	if c.zw != nil {
+		if err := c.zw.Close(); err != nil {
+			return err
+		}
+	}
+
+	return c.rawConn.Close()
+}
+
+func (c *zlibConn) LocalAddr() net.Addr {
+	return c.rawConn.LocalAddr()
+}
+
+func (c *zlibConn) RemoteAddr() net.Addr {
+	return c.rawConn.RemoteAddr()
+}
+
+func (c *zlibConn) SetDeadline(t time.Time) error {
+	return c.rawConn.SetDeadline(t)
+}
+
+func (c *zlibConn) SetReadDeadline(t time.Time) error {
+	return c.rawConn.SetReadDeadline(t)
+}
+
+func (c *zlibConn) SetWriteDeadline(t time.Time) error {
+	return c.rawConn.SetWriteDeadline(t)
+}
+
 var _ Transport = new(sslsocket)
 
-func NewSSLSocket(hostPort string, tlsConfig *tls.Config) (Transport, error) {
+func NewSSLSocket(hostPort string, tlsConfig *tls.Config, compressionLevel int) (Transport, error) {
+
 	addr, err := net.ResolveTCPAddr("tcp", hostPort)
 	if err != nil {
 		return nil, err
 	}
 
 	s := &sslsocket{}
-	s.Conn, err = tls.Dial(addr.Network(), addr.String(), tlsConfig)
-	if err != nil {
-		return nil, fmt.Errorf("TLS connection failed: %w", err)
+
+	if compressionLevel > 0 {
+		rawConn, err := net.Dial(addr.Network(), addr.String())
```

Comment:
Now for me it seems that we don't need SSLSocket at all but need a new transport - TLS. Then the client pseudo code may look like this:
```
transport.NewSocket
if buffer {
    transport.NewBuffered
}
if compress {
    transport.NewZlib
}
if TLS {
    transport.NewTLS
}
```
And inside NewTLS we may have the following:
```

		conn := tls.Client(zlibConn, tlsConfig)
		if err != nil {
			return nil, err
		}

		err = conn.Handshake()
		if err != nil {
			return nil, err
		}
```
And I believe that we may use the same Zlib transport for both cases - the current one should be just adapted a bit for working with both tls and plain connections.

Reply:
Good idea, I didn't notice that the current ZLib transport almost implemented the net.Conn interface.

Reply:
I'm trying to implement it with the current transports, however the TLS transport requires flushes on writes otherwise the connection gets terminated during handshake, if I modify the zlib transport and buffered transport to flush on writes unit tests starts to fail. 

I'm not sure how I would re-use the current transports and simultaneously make them compatible with the TLS. 

From what I can understand of the buffered transport it is special to this library (not part of the Avro protocol) to send huge messages with buffering. I think that should be added last to the transport chain instead of directly on top of the tcp socket, that way the buffered transport won't need to flush on writes or am I missing something?

Reply:
Then maybe we can allow enabling buffered transport only for plain socket. If someone tries to enable TLS and buffering we may raise and error. In this case I'd add a new ticket to support buffering and TLS at the same time with detailed description of what is wrong with this now. Then add a link to this ticket right inside the related code. For now it will be a solution especially if compression and tls work fine together.

Reply:
I'd like to be more helpful here but unfortunately have no time to dive deeper in this issue.

Reply:
Ah, don't worry.

I've actually got everything working when I re-ordered the transports, with all combinations against a real flume server, however 2 tests are currently failing. 

```--- FAIL: TestZlibTransport_Write (0.00s)
    --- FAIL: TestZlibTransport_Write/short_write (0.00s)
        zlib_test.go:81: 
                Error Trace:    /Users/callegustafsson/Source/avroipc/transports/zlib_test.go:81
                Error:          Not equal: 
                                expected: []byte{0x78, 0x1}
                                actual  : []byte{0x78, 0x1, 0x0, 0x4, 0x0, 0xfb, 0xff, 0x74, 0x65, 0x73, 0x74, 0x0, 0x0, 0x0, 0xff, 0xff}
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,3 +1,3 @@
                                -([]uint8) (len=2) {
                                - 00000000  78 01                                             |x.|
                                +([]uint8) (len=16) {
                                + 00000000  78 01 00 04 00 fb ff 74  65 73 74 00 00 00 ff ff  |x......test.....|
                                 }
                Test:           TestZlibTransport_Write/short_write
    --- FAIL: TestZlibTransport_Write/with_close (0.00s)
        zlib_test.go:94: 
                Error Trace:    /Users/callegustafsson/Source/avroipc/transports/zlib_test.go:94
                Error:          Not equal: 
                                expected: []byte{0x78, 0x1, 0x0, 0x4, 0x0, 0xfb, 0xff, 0x74, 0x65, 0x73, 0x74, 0x1, 0x0, 0x0, 0xff, 0xff, 0x4, 0x5d, 0x1, 0xc1}
                                actual  : []byte{0x78, 0x1, 0x0, 0x4, 0x0, 0xfb, 0xff, 0x74, 0x65, 0x73, 0x74, 0x0, 0x0, 0x0, 0xff, 0xff, 0x1, 0x0, 0x0, 0xff, 0xff, 0x4, 0x5d, 0x1, 0xc1}
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,4 +1,4 @@
                                -([]uint8) (len=20) {
                                - 00000000  78 01 00 04 00 fb ff 74  65 73 74 01 00 00 ff ff  |x......test.....|
                                - 00000010  04 5d 01 c1                                       |.]..|
                                +([]uint8) (len=25) {
                                + 00000000  78 01 00 04 00 fb ff 74  65 73 74 00 00 00 ff ff  |x......test.....|
                                + 00000010  01 00 00 ff ff 04 5d 01  c1                       |......]..|
                                 }
                Test:           TestZlibTransport_Write/with_close
FAIL
FAIL    github.com/myzhan/avroipc/transports    1.120s
FAIL```

This is caused by every write to the zlib transport flushes to be able to be used with the TLS transport. which should be fine. I'm going to change those tests to adhere to the new data